src/shims/sync.rs

```diff
@@ -345,9 +345,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let kind = this.read_scalar(kind_op)?.not_undef()?;
         if kind == this.eval_libc("PTHREAD_MUTEX_NORMAL")? {
             // In `glibc` implementation, the numeric values of
-            // `PTHREAD_MUTEX_NORMAL` and `PTHREAD_MUTEX_DEFAULT` are equal, but
-            // they have different behaviour in some cases. Therefore, we assume
-            // that `pthread_mutexattr_settype` is ever called only with
+            // `PTHREAD_MUTEX_NORMAL` and `PTHREAD_MUTEX_DEFAULT` are equal.
+            // However, a mutex created by explicitly passing
+            // `PTHREAD_MUTEX_NORMAL` type has in some cases different behaviour
+            // from the default mutex for which the type was not explicitly
+            // specified. For a more detailed discussion, please see
+            // https://github.com/rust-lang/miri/issues/1419#.
+            //
+            // To distinguish these two cases, we do the same trick as glibc: we
+            // assume that `pthread_mutexattr_settype` is ever called only with
```

Comment:
"assume" here is the wrong term IMO... we don't make any assumption about client behavior. We "just" distinguish "NORMAL/DEFAULT was explicitly set" from "the value was not changed (left at its default value)".